```{r}
#saveRDS(ALL.combined2, file = "ALL_combined_seurat_object.rds")
ALL.combined2 <- readRDS("ALL_combined_seurat_object.rds")
library(dplyr)
library(Seurat)
library(Seurat)
library(patchwork)
ElbowPlot(ALL.combined2)
#DimPlot(ALL.combined2, reduction ="umap", label = TRUE)
## Using presto 
#install.packages("devtools")
#devtools::install_github("immunogenomics/presto")
#wilcoxauc(X, y)
#wilcoxauc(seurat_object, 'group_name')
#wilcoxauc(sce_object, 'group_name')
# Step 5: Perform clustering
library(dplyr)
library(Seurat)
#ALL.combined2 <- FindNeighbors(ALL.combined2, reduction = "umap", dims = 1:20)
ALL.combined2 <- FindClusters(ALL.combined2, resolution = 0.1) # Adjust resolution as needed
### presto 
# Step 6: Visualize the UMAP with clusters
#View(ALL.combined$meta.data)
ALL.combined2 <- RunUMAP(ALL.combined2, dims = 1:10, min.dist = 0.1, n.neighbors = 30)
DimPlot(ALL.combined2, reduction = 'umap', label = TRUE,raster = FALSE)
DimPlot(ALL.combined2, reduction = 'umap', label = TRUE,raster = FALSE,split.by = "orig.ident")

##############
# Set the default assay to RNA (or another assay where FOXP3 is located)
DefaultAssay(ALL.combined2) <- "RNA"  # or "SCT" or "integrated", depending on your workflow
saveRDS(ALL.combined2, file = "ALL_combined_seurat_object.rds")
ALL.combined2 <- readRDS("ALL_combined_seurat_object.rds")
DimPlot(ALL.combined2, reduction = 'umap', label = TRUE,raster = FALSE)
DimPlot(ALL.combined2, reduction = 'umap', label = TRUE,raster = FALSE,split.by = "orig.ident")

```

```{r}
library(ggplot2)

library(Seurat)
library(dplyr)
library(Seurat)
library(dplyr)

# Load necessary libraries
library(dplyr)
library(ggplot2)
library(tidyr)
# Load necessary libraries
library(dplyr)
library(ggplot2)
library(tidyr)
library(dplyr)
library(ggplot2)
library(tidyr)

# Join data layers for the assay "RNA"
ALL.combined2 <- JoinLayers(ALL.combined2, assay = "RNA")

# Find markers for cluster 0
DefaultAssay(ALL.combined2) <- "RNA"
cluster0 <- FindConservedMarkers(ALL.combined2, ident.1 = 0, only.pos = TRUE,grouping.var = "orig.ident", verbose = FALSE)
VlnPlot(ALL.combined2, features = "GZMK", group.by = "celltypes",pt.size = 0)

#DotPlot(ALL.combined2, features = c("PPP1R2C", "GZMK", "CRTAM", "TRGV10", "ENSG00000235785", "GREM2", #"IL2", "CD8A"))
# Find marker for cluster 1
cluster1 <- FindConservedMarkers(ALL.combined2, ident.1 = 1, only.pos = TRUE,grouping.var = "orig.ident", verbose = FALSE)
VlnPlot(ALL.combined2, features = "BANK1", group.by = "celltypes",pt.size = 0)

DotPlot(ALL.combined2, features = c("GALNTL6", "CLEC17A", "FCER2", "BANK1", "FCRL2", "NIBAN3", "ENSG00000289963", "CR2"))
# Find marker for cluster 2
cluster2 <- FindConservedMarkers(ALL.combined2, ident.1 = 2, only.pos = TRUE,grouping.var = "orig.ident", verbose = FALSE)
VlnPlot(ALL.combined2, features = "CFTR", group.by = "celltypes",pt.size = 0)
VlnPlot(ALL.combined2, features = "CYP2C18", group.by = "celltypes",pt.size = 0)


DotPlot(ALL.combined2, features = c("CYP2C19", "WNT7B", "TM4SF5", "ENTPD8", "CYP2C18", "IYD"))
# find marker for cluster 3
cluster3 <- FindConservedMarkers(ALL.combined2, ident.1 = 3, only.pos = TRUE,grouping.var = "orig.ident", verbose = FALSE)
VlnPlot(ALL.combined2, features = "LUM", group.by = "celltypes",pt.size = 0)

DotPlot(ALL.combined2, features = c("FAM180A", "BARX1", "LINC01638", "OSR1", "WT1-AS", "PDGFRA", "TAC1", "PLA2G2A", "LUM", "NPTX1"))
# find marker for cluster 4
cluster4 <- FindConservedMarkers(ALL.combined2, ident.1 = 4, only.pos = TRUE,grouping.var = "orig.ident", verbose = FALSE)
VlnPlot(ALL.combined2, features = "PRSS3", group.by = "celltypes",pt.size = 0)

DotPlot(ALL.combined2, features = c("NDUFA4L2", "FOXS1", "KCNK17", "RGS5", "CSPG4", "IGHG1"))
# find marker for cluster 5
cluster5 <- FindConservedMarkers(ALL.combined2, ident.1 = 5, only.pos = TRUE,grouping.var = "orig.ident", verbose = FALSE)
VlnPlot(ALL.combined2, features = "C1QA", group.by = "celltypes",pt.size = 0)

DotPlot(ALL.combined2, features = c("LILRB5", "C1QA", "C1QC",  "MARCO", "SIGLEC1", "LINC03070", "C1QB", "CD163"))
# find marker for cluster 6
cluster6 <- FindConservedMarkers(ALL.combined2, ident.1 = 6, only.pos = TRUE,grouping.var = "orig.ident", verbose = FALSE)
VlnPlot(ALL.combined2, features = "RTKN2", group.by = "celltypes",pt.size = 0)

DotPlot(ALL.combined2, features = c( "FOXP3", "ENSG00000263655", "RTKN2", "LINC01281", "ENSG00000224610", "TNFRSF4"))
# find marker for cluster 7
cluster7 <- FindConservedMarkers(ALL.combined2, ident.1 = 7, only.pos = TRUE,grouping.var = "orig.ident", verbose = FALSE)
VlnPlot(ALL.combined2, features = "TEK", group.by = "celltypes",pt.size = 0)

DotPlot(ALL.combined2, features = c("ANO2", "TPO", "MYCN", "KLHL4", "RXFP1", "SLC5A4", "TEK"))
# find marker for cluster 8
cluster8 <- FindConservedMarkers(ALL.combined2, ident.1 = 8, only.pos = TRUE,grouping.var = "orig.ident", verbose = FALSE)
VlnPlot(ALL.combined2, features = "FCGR3B", group.by = "celltypes",pt.size = 0)

DotPlot(ALL.combined2, features = c("FCGR3B","ENSG00000289573", "ENSG00000255801", "CXCR1", "CMTM2", "LINC00664", "ENSG00000230492", "PHOSPHO1", "CXCR2", "CEACAM3", "FPR2"))
# find marker for cluster 9
cluster9 <- FindConservedMarkers(ALL.combined2, ident.1 = 9, only.pos = TRUE,grouping.var = "orig.ident", verbose = FALSE)
VlnPlot(ALL.combined2, features = "THEMIS", group.by = "celltypes",pt.size = 0)

DotPlot(ALL.combined2, features = c("GML", "FCRL6", "CCR2", "CDC42EP3-AS1", "PDXP-DT", "PTGDR", "IRGM", "LINC00861", "ENSG00000236266", "GRAP2", "SAMD3"))
# find marker for cluster 10
cluster10 <- FindConservedMarkers(ALL.combined2, ident.1 = 10, only.pos = TRUE,grouping.var = "orig.ident", verbose = FALSE)
VlnPlot(ALL.combined2, features = "GZMB", group.by = "celltypes",pt.size = 0)

DotPlot(ALL.combined2, features = c("FGFBP2","GNLY", "KLRF1", "MYOM2", "NCR1","B3GAT1", "CX3CR1", "NME8", "GZMB", "TRDC"))
# find marker for cluster 11
cluster11 <- FindConservedMarkers(ALL.combined2, ident.1 = 11, only.pos = TRUE,grouping.var = "orig.ident", verbose = FALSE)
VlnPlot(ALL.combined2, features = "TPSB2", group.by = "celltypes",pt.size = 0)

DotPlot(ALL.combined2, features = c("TPSB2", "TPSAB1", "CTSG", "MS4A2", "SLC18A2-AS1", "CPA3", "HDC", "ADCYAP1", "GCSAML", "IL1RL1", "CDK15", "SLC8A3", "CALB2"))
# find marker for cluster 12
cluster12 <- FindConservedMarkers(ALL.combined2, ident.1 = 12, only.pos = TRUE,grouping.var = "orig.ident", verbose = FALSE)
VlnPlot(ALL.combined2, features = "NKAIN3", group.by = "celltypes",pt.size = 0)

DotPlot(ALL.combined2, features = c("SOX2", "PRIMA1", "GAP43", "FOXD3", "PLP1", "ABCB5","FGF5", "NDP", "SYT10", "CDH19", "GDNF", "NKAIN3", "KIRREL3"))
# find marker for cluster 13
cluster13 <- FindConservedMarkers(ALL.combined2, ident.1 = 13, only.pos = TRUE,grouping.var = "orig.ident", verbose = FALSE)
VlnPlot(ALL.combined2, features = "PRPSAP2", group.by = "celltypes",pt.size = 0)
#NIBAN3
DotPlot(ALL.combined2, features = c("PRPSAP2", "CCDC144BP","AICDA", "ELL3", "HRK", "SLC2A5","HTR3A", "TCL1A", "CFAP299"))
FeaturePlot(ALL.combined2, features = c('MYT1L'))
# find marker for cluster 14
cluster14 <- FindConservedMarkers(ALL.combined2, ident.1 = 14, only.pos = TRUE,grouping.var = "orig.ident", verbose = FALSE)
VlnPlot(ALL.combined2, features = "IL4I1", group.by = "celltypes",pt.size = 0)

DotPlot(ALL.combined2, features = c("LINC00299", "IL4I1", "IL23R", "TNFRSF18", "KRT86", "SCX", "SCN1B", "TNFSF11", "HPN", "TRD-AS1", "HTR1F"))
# find marker for cluster 15
cluster15 <- FindConservedMarkers(ALL.combined2, ident.1 = 15, only.pos = TRUE,grouping.var = "orig.ident", verbose = FALSE)
VlnPlot(ALL.combined2, features = "SCG5", group.by = "celltypes",pt.size = 0)

DotPlot(ALL.combined2, features = c("CACNA1A", "SCG5", "CHGB", "KCNMB2", "TTR", "PCSK1N", "RIMS2", "HTR1F", "SCGN", "TMEM132D", "DPP6", "HS6ST3", "SNAP25", "PCSK2","ILDR2"))

```

```{r}
library(patchwork)
# Load required libraries
library(Seurat)
library(ggplot2)
library(Seurat)
library(Seurat)
library(dplyr)

# Assuming ALL.combined2 is your Seurat object

# Initialize an empty list to store top genes for all clusters
top_genes <- list()

# Loop through clusters 0 to 15
for (i in 0:15) {
  # Find conserved markers for the current cluster
  conserved_markers <- FindConservedMarkers(ALL.combined2, ident.1 = i, 
                                             only.pos = TRUE, 
                                             grouping.var = "orig.ident", 
                                             verbose = FALSE)
  
  # Select top 5 markers and store in the list
  top_genes[[i + 1]] <- head(rownames(conserved_markers), 3)
}

# Combine all top genes into a single vector
all_top_genes <- unique(unlist(top_genes))

# Create the DotPlot using the combined top genes
dot_plot <- DotPlot(ALL.combined2, features = all_top_genes) +
  labs(title = "Dot Plot of Top 5 Cell Type-Specific Genes Across Clusters") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Display the DotPlot
print(dot_plot)

# Save the DotPlot (optional)
ggsave("combined_dot_plot.png", plot = dot_plot, width = 10, height = 8)

```






```{r}
View(ALL.combined2@meta.data)
Idents(ALL.combined2)
ALL.combined2$celltypes<-Idents(ALL.combined2)

ALL.combined2$norpdac_celltype <- paste(ALL.combined2$orig.ident,sep = "_",Idents(ALL.combined2))
Idents(ALL.combined2) <-"norpdac_celltype"
DimPlot(ALL.combined2, reduction = 'umap', label = TRUE, split.by = "orig.ident")

DimPlot(ALL.combined2, reduction = 'umap', label = FALSE)
#DimPlot(ALL.combined2, reduction = 'umap', label = TRUE, .by = "orig.ident")

```

```{r}

ALL.combined2 <- RenameIdents(ALL.combined2,'0'= 'CD8T cell')
ALL.combined2 <- RenameIdents(ALL.combined2,'1'= 'naive B cell')
ALL.combined2 <- RenameIdents(ALL.combined2,'2'= 'Epithelial cell')
ALL.combined2 <- RenameIdents(ALL.combined2,'3'= 'Fibroblast cell')
ALL.combined2 <- RenameIdents(ALL.combined2,'4'= 'SMC')
ALL.combined2 <- RenameIdents(ALL.combined2,'5'= 'Macrophages')
ALL.combined2 <- RenameIdents(ALL.combined2,'6'= 't-reg')
ALL.combined2 <- RenameIdents(ALL.combined2,'7'= 'Endothelial cell')
ALL.combined2 <- RenameIdents(ALL.combined2,'8'= 'Neutrophil ')
ALL.combined2 <- RenameIdents(ALL.combined2,'9'= 'CD4T cell')
ALL.combined2 <- RenameIdents(ALL.combined2,'10'= 'NK cell')
ALL.combined2 <- RenameIdents(ALL.combined2,'11'= 'Basophill cell')
ALL.combined2 <- RenameIdents(ALL.combined2,'12'= 'Oligodendrocytes cell')
ALL.combined2 <- RenameIdents(ALL.combined2,'13'= 'B cell')
ALL.combined2 <- RenameIdents(ALL.combined2,'14'= 'naive T cell')
ALL.combined2 <- RenameIdents(ALL.combined2,'15'= 'Endocrine cell')
#DimPlot(ALL.combined2, reduction = 'umap', label = TRUE, raster = FALSE)
DimPlot(ALL.combined2, reduction = 'umap', label = TRUE,raster = FALSE,split.by = "orig.ident")
#cluater0_469.markers <- FindMarkers(ALL.combined2, ident.1 = 0,ident.2 = c(4,6,9),min.pct = 0.25)
```


```{r}
# Set the default assay to RNA to ensure we analyze gene expression
DefaultAssay(ALL.combined2) <- "RNA"
ALL.combined2 <- JoinLayers(ALL.combined2, assay = "RNA")



```


```{r}
FeaturePlot(ALL.combined2, reduction = 'umap', 
            features = c('KRT17'),
            split.by = 'orig.ident', min.cutoff = 'q10')
```

```{r}
# to see the celltype variation by conditions 
FeaturePlot(ALL.combined2, reduction = 'umap', 
            features = c('FGFBP2'),
            split.by = 'orig.ident', min.cutoff = 'q10')
```

```{r}
library(Seurat)
library(dplyr)

# Assuming ALL.combined2 is your Seurat object

```



```{r}
# Assuming 'celltype' includes both condition and cluster information
Idents(ALL.combined2) <- "norpdac_celltype"  # Adjust if necessary

## DEG for cluster 0
cluster0_deg <- FindMarkers(ALL.combined2,ident.1 = 'PDAC_CD8T cell', ident.2 = 'NOR_CD8T cell',only.pos = TRUE, verbose = FALSE)
Idents(ALL.combined2) <- "norpdac_celltype"  # Adjust if necessary
top20cluster0_genes <- head(cluster0_deg[order(cluster0_deg$avg_log2FC, decreasing = TRUE), ], 20)

## DEG for cluster 1
cluster1_deg <- FindMarkers(ALL.combined2,ident.1 = "PDAC_naive B cell", ident.2 = "NOR_naive B cell",only.pos = TRUE, verbose = FALSE)
top20cluster1_genes <- head(cluster1_deg[order(cluster1_deg$avg_log2FC, decreasing = TRUE), ], 20)
VlnPlot(ALL.combined2, features = "CLDN18", split.by = "orig.ident", group.by = "celltypes",pt.size = 0,combine = FALSE)
FeaturePlot(ALL.combined2, reduction = 'umap', 
            features = c('CREM'),
            split.by = 'orig.ident', min.cutoff = 'q10')
## DEG for cluster 2
Idents(ALL.combined2) <- "norpdac_celltype"  # Adjust if necessary

cluster2_deg <- FindMarkers(ALL.combined2,ident.1 = "PDAC_Epithelial cell", ident.2 = "NOR_Epithelial cell",only.pos = TRUE, verbose = FALSE)
top20cluster2_genes <- head(cluster2_deg[order(cluster2_deg$avg_log2FC, decreasing = TRUE), ], 20)
## DEG for cluster 3

cluster3_deg <- FindMarkers(ALL.combined2,ident.1 = 'PDAC_Fibroblast cell', ident.2 = 'NOR_Fibroblast cell',only.pos = TRUE, verbose = FALSE)
top20cluster3_genes <- head(cluster3_deg[order(cluster3_deg$avg_log2FC, decreasing = TRUE), ], 20)
## DEG for cluster 4
Idents(ALL.combined2) <- "norpdac_celltype"  # Adjust if necessary

cluster4_deg <- FindMarkers(ALL.combined2,ident.1 = "PDAC_SMC", ident.2 = "NOR_SMC",only.pos = TRUE, verbose = FALSE)
top20cluster4_genes <- head(cluster4_deg[order(cluster4_deg$avg_log2FC, decreasing = TRUE), ], 20)
## DEG for cluster 5
Idents(ALL.combined2) <- "norpdac_celltype"  # Adjust if necessary

cluster5_deg <- FindMarkers(ALL.combined2,ident.1 = "PDAC_Macrophages", ident.2 = "NOR_Macrophages",only.pos = TRUE, verbose = FALSE)
top20cluster5_genes <- head(cluster5_deg[order(cluster5_deg$avg_log2FC, decreasing = TRUE), ], 20)
## DEG for cluster 6
Idents(ALL.combined2) <- "norpdac_celltype"  # Adjust if necessary

cluster6_deg <- FindMarkers(ALL.combined2,ident.1 = "PDAC_t-reg", ident.2 = "NOR_t-reg",only.pos = TRUE, verbose = FALSE)
top20cluster6_genes <- head(cluster6_deg[order(cluster6_deg$avg_log2FC, decreasing = TRUE), ], 20)
## DEG for cluster 7
Idents(ALL.combined2) <- "norpdac_celltype"  # Adjust if necessary
cluster7_deg <- FindMarkers(ALL.combined2,ident.1 = "PDAC_Endothelial cell", ident.2 = "NOR_Endothelial cell",only.pos = TRUE, verbose = FALSE)
top20cluster7_genes <- head(cluster7_deg[order(cluster7_deg$avg_log2FC, decreasing = TRUE), ], 20)
## DEG for cluster 8
Idents(ALL.combined2) <- "norpdac_celltype"  # Adjust if necessary

cluster8_deg <- FindMarkers(ALL.combined2,ident.1 = "PDAC_Neutrophil " , ident.2 = "NOR_Neutrophil ",only.pos = TRUE, verbose = FALSE)
top20cluster8_genes <- head(cluster8_deg[order(cluster8_deg$avg_log2FC, decreasing = TRUE), ], 20)
## DEG for cluster 9
Idents(ALL.combined2) <- "norpdac_celltype"  # Adjust if necessary

cluster9_deg <- FindMarkers(ALL.combined2,ident.1 = "PDAC_CD4T cell", ident.2 = "NOR_CD4T cell",only.pos = TRUE, verbose = FALSE)
top20cluster9_genes <- head(cluster9_deg[order(cluster9_deg$avg_log2FC, decreasing = TRUE), ], 20)
## DEG for cluster 10
Idents(ALL.combined2) <- "norpdac_celltype"  # Adjust if necessary

cluster10_deg <- FindMarkers(ALL.combined2,ident.1 = "PDAC_NK cell", ident.2 = "NOR_NK cell",only.pos = TRUE, verbose = FALSE)
top20cluster10_genes <- head(cluster10_deg[order(cluster10_deg$avg_log2FC, decreasing = TRUE), ], 20)
## DEG for cluster 11
Idents(ALL.combined2) <- "norpdac_celltype"  # Adjust if necessary

cluster11_deg <- FindMarkers(ALL.combined2,ident.1 = "PDAC_Basophill cell", ident.2 = "NOR_Basophill cell",only.pos = TRUE, verbose = FALSE)
top20cluster11_genes <- head(cluster11_deg[order(cluster11_deg$avg_log2FC, decreasing = TRUE), ], 20)
## DEG for cluster 12
Idents(ALL.combined2) <- "norpdac_celltype"  # Adjust if necessary

cluster12_deg <- FindMarkers(ALL.combined2,ident.1 = "PDAC_Oligodendrocytes cell", ident.2 = "NOR_Oligodendrocytes cell",only.pos = TRUE, verbose = FALSE)
top20cluster12_genes <- head(cluster12_deg[order(cluster12_deg$avg_log2FC, decreasing = TRUE), ], 20)
## DEG for cluster 13
Idents(ALL.combined2) <- "norpdac_celltype"  # Adjust if necessary

cluster13_deg <- FindMarkers(ALL.combined2,ident.1 = "PDAC_B cell", ident.2 = "NOR_B cell",only.pos = TRUE, verbose = FALSE)
top20cluster13_genes <- head(cluster13_deg[order(cluster13_deg$avg_log2FC, decreasing = TRUE), ], 20)
## DEG for cluster 14
Idents(ALL.combined2) <- "norpdac_celltype"  # Adjust if necessary

cluster14_deg <- FindMarkers(ALL.combined2,ident.1 = "PDAC_naive T cell", ident.2 = "NOR_naive T cell",only.pos = TRUE, verbose = FALSE)
top20cluster14_genes <- head(cluster14_deg[order(cluster14_deg$avg_log2FC, decreasing = TRUE), ], 20)
## DEG for cluster 15
Idents(ALL.combined2) <- "norpdac_celltype"  # Adjust if necessary

cluster15_deg <- FindMarkers(ALL.combined2,ident.1 = "PDAC_Endocrine cell", ident.2 = "NOR_Endocrine cell",only.pos = TRUE, verbose = FALSE)
top20cluster15_genes <- head(cluster15_deg[order(cluster15_deg$avg_log2FC, decreasing = TRUE), ], 20)
```
```{r}
# List of all DEG data frames
cluster_DEGs <- list(cluster0_deg, cluster1_deg, cluster2_deg, cluster3_deg, cluster4_deg,
                     cluster5_deg, cluster6_deg, cluster7_deg, cluster8_deg, cluster9_deg,
                     cluster10_deg, cluster11_deg, cluster12_deg, cluster13_deg, cluster14_deg,
                     cluster15_deg)

# Select top 10 DEGs per cluster based on avg_log2FC and combine them
top_genes <- unique(unlist(lapply(cluster_DEGs, function(deg) {
  rownames(head(deg[order(deg$avg_log2FC, decreasing = TRUE), ], 10))
})))
#
# Set the identity of cells in Seurat object to 'seurat_clusters' (or the appropriate grouping)
Idents(ALL.combined2) <- "seurat_clusters"

# Plot the heatmap for the selected DEGs
DoHeatmap(
  ALL.combined2,
  features = top_genes,
  group.by = "seurat_clusters"
) + scale_fill_gradientn(colors = c("blue", "white", "red"))
##############
# Extract normalized data and scale by rows (genes)
mat <- GetAssayData(ALL.combined2[["RNA"]], slot = "data")[top_genes, ] %>% as.matrix()
mat <- t(scale(t(mat)))

# Check the value range
quantile(mat, c(0.1, 0.95))
#
# Define the color scale function
# Install circlize and ComplexHeatmap packages
install.packages("circlize")
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
}
BiocManager::install("ComplexHeatmap")

library(circlize)
col_fun <- colorRamp2(c(-2, 0, 2), c("blue", "white", "red"))
#####
# Load ComplexHeatmap package
library(ComplexHeatmap)

# Extract cluster annotations from metadata
cluster_anno <- ALL.combined2@meta.data$seurat_clusters

# Plot the heatmap with ComplexHeatmap
Heatmap(
  mat,
  name = "Expression",
  col = col_fun,
  show_row_names = TRUE,
  show_column_names = FALSE,
  cluster_rows = TRUE,
  cluster_columns = TRUE,
  column_split = cluster_anno,  # Split columns by clusters
  row_names_gp = gpar(fontsize = 6),  # Adjust row label font size for clarity
  top_annotation = HeatmapAnnotation(
    cluster = cluster_anno,
    col = list(cluster = Seurat::PurpleAndYellow())
  ),
  heatmap_legend_param = list(
    title = "Scaled Expression",
    title_position = "topcenter",
    legend_direction = "horizontal"
  )
)
# Get unique clusters and assign colors
unique_clusters <- unique(cluster_anno)
cluster_colors <- setNames(Seurat::PurpleAndYellow(length(unique_clusters)), unique_clusters)
# Define the heatmap with ComplexHeatmap and annotations
Heatmap(
  mat,
  name = "Expression",
  col = col_fun,
  show_row_names = TRUE,
  show_column_names = FALSE,
  cluster_rows = TRUE,
  cluster_columns = TRUE,
  column_split = cluster_anno,  # Split columns by clusters
  row_names_gp = gpar(fontsize = 6),  # Adjust row label font size for clarity
  top_annotation = HeatmapAnnotation(
    cluster = cluster_anno,
    col = list(cluster = cluster_colors)  # Use named vector for cluster colors
  ),
  heatmap_legend_param = list(
    title = "Scaled Expression",
    title_position = "topcenter",
    legend_direction = "horizontal"
  )
)

```

```{r}
# List of all DEG data frames
cluster_DEGs <- list(cluster0_deg, cluster1_deg, cluster2_deg, cluster3_deg, cluster4_deg,
                     cluster5_deg, cluster6_deg, cluster7_deg, cluster8_deg, cluster9_deg,
                     cluster10_deg, cluster11_deg, cluster12_deg, cluster13_deg, cluster14_deg,
                     cluster15_deg)

# Get the top 10 DEGs for each cluster and create a matrix
top_genes <- unique(unlist(lapply(cluster_DEGs, function(deg) {
  rownames(head(deg[order(deg$avg_log2FC, decreasing = TRUE), ], 10))
})))

# Extract the expression matrix for the top genes
mat <- GetAssayData(ALL.combined2[["RNA"]], slot = "data")[top_genes, ] %>% as.matrix()
mat <- t(scale(t(mat)))  # Scale the expression matrix

# Load necessary libraries
library(ComplexHeatmap)
library(circlize)

# Define color function
col_fun <- colorRamp2(c(-2, 0, 2), c("blue", "white", "red"))

# Define a function to create a color vector for highlighting top genes
highlight_colors <- rep("white", length(top_genes))  # Default to white
highlight_colors[which(top_genes %in% unlist(lapply(cluster_DEGs, function(deg) {
  head(rownames(deg[order(deg$avg_log2FC, decreasing = TRUE), ]), 10)
}))) ] <- "yellow"  # Highlight top genes in yellow
# Plot the heatmap
# Load necessary libraries
library(ComplexHeatmap)
library(circlize)

# Extract the top DEGs and scale the expression matrix
top_genes <- unique(unlist(lapply(cluster_DEGs, function(deg) {
  rownames(head(deg[order(deg$avg_log2FC, decreasing = TRUE), ], 10))
})))

# Extract the expression matrix for the top genes
mat <- GetAssayData(ALL.combined2[["RNA"]], slot = "data")[top_genes, ] %>% as.matrix()
mat <- t(scale(t(mat)))  # Scale the expression matrix

# Define color function for heatmap
col_fun <- colorRamp2(c(-2, 0, 2), c("blue", "white", "red"))

# Get the cluster annotations
cluster_anno <- ALL.combined2@meta.data$seurat_clusters

# Create a vector for row label colors, highlighting top genes
row_label_colors <- ifelse(top_genes %in% unlist(lapply(cluster_DEGs, function(deg) {
  head(rownames(deg[order(deg$avg_log2FC, decreasing = TRUE), ], 10))
})), "yellow", "black")

# Plot the heatmap
heatmap_plot <- Heatmap(
  mat,
  name = "Expression",
  col = col_fun,
  show_row_names = TRUE,
  show_column_names = FALSE,
  cluster_rows = TRUE,
  cluster_columns = TRUE,
  row_names_gp = gpar(fontsize = 8, col = row_label_colors),  # Set row name colors
  column_split = cluster_anno,  # Split columns by clusters
  top_annotation = HeatmapAnnotation(
    cluster = cluster_anno,
    col = list(cluster = Seurat::PurpleAndYellow())
  ),
  heatmap_legend_param = list(
    title = "Scaled Expression",
    title_position = "topcenter",
    legend_direction = "horizontal"
  )
)

# Draw the heatmap
draw(heatmap_plot)


# Load necessary libraries
library(ComplexHeatmap)
library(circlize)
library(Seurat)

# Extract the top DEGs and scale the expression matrix
top_genes <- unique(unlist(lapply(cluster_DEGs, function(deg) {
  rownames(head(deg[order(deg$avg_log2FC, decreasing = TRUE), ], 10))
})))

# Extract the expression matrix for the top genes
mat <- GetAssayData(ALL.combined2[["RNA"]], slot = "data")[top_genes, ] %>% as.matrix()
mat <- t(scale(t(mat)))  # Scale the expression matrix

# Define color function for heatmap
col_fun <- colorRamp2(c(-2, 0, 2), c("blue", "white", "red"))

# Get the cluster annotations
cluster_anno <- ALL.combined2@meta.data$seurat_clusters

# Create a vector for row label colors, highlighting top genes
row_label_colors <- ifelse(top_genes %in% unlist(lapply(cluster_DEGs, function(deg) {
  head(rownames(deg[order(deg$avg_log2FC, decreasing = TRUE), ], 10))
})), "yellow", "black")

# Define colors for the clusters
unique_clusters <- unique(cluster_anno)
cluster_colors <- setNames(Seurat::PurpleAndYellow(length(unique_clusters)), unique_clusters)

# Plot the heatmap
heatmap_plot <- Heatmap(
  mat,
  name = "Expression",
  col = col_fun,
  show_row_names = TRUE,
  show_column_names = FALSE,
  cluster_rows = TRUE,
  cluster_columns = TRUE,
  row_names_gp = gpar(fontsize = 8, col = row_label_colors),  # Set row name colors
  column_split = cluster_anno,  # Split columns by clusters
  top_annotation = HeatmapAnnotation(
    cluster = cluster_anno,
    col = list(cluster = cluster_colors)  # Use the named vector of colors
  ),
  heatmap_legend_param = list(
    title = "Scaled Expression",
    title_position = "topcenter",
    legend_direction = "horizontal"
  )
)

# Draw the heatmap
draw(heatmap_plot)

```

```{r}
## cluster 0
top20cluster0_genes$celltype <- "CD8T cell"
View(top20cluster0_genes)
## cluster 1
top20cluster1_genes$celltype <- "naive B cell"
View(top20cluster1_genes)
## cluster 2
top20cluster2_genes$celltype <- "Epithelial cell"
View(top20cluster2_genes)
## cluster 3
top20cluster3_genes$celltype <- "Fibroblast cell"
View(top20cluster3_genes)
## cluster4
top20cluster4_genes$celltype <- "SMC"
View(top20cluster4_genes)
## cluster 5
top20cluster5_genes$celltype <- "Macrophages"
View(top20cluster5_genes)
## cluster 6
top20cluster6_genes$celltype <- "t-reg"
View(top20cluster6_genes)
## cluster 7
top20cluster7_genes$celltype <- "Endothelial cell"
View(top20cluster7_genes)
## cluster 8
top20cluster8_genes$celltype <- "Neutrophil "
View(top20cluster8_genes)
## cluster 9
top20cluster9_genes$celltype <- "CD4T cell"
View(top20cluster9_genes)
## cluster 10
top20cluster10_genes$celltype <- "NK cell"
View(top20cluster10_deg)
## cluster11 
top20cluster11_genes$celltype <- "Basophill cell"
View(top20cluster11_deg)
## cluster 12 
top20cluster12_genes$celltype <- "Oligodendrocytes cell"
View(top20cluster12_genes)
## cluster 13
top20cluster13_genes$celltype <- "B cell"
View(top20cluster13_genes)
## cluster 14
top20cluster14_genes$celltype <- "naive T cell"
View(top20cluster14_genes)
## cluster 15 
top20cluster15_genes$celltype <- "Endocrine cell"
View(top20cluster15_genes)



```

## cluster```{r}
# Get the cluster identities of all cells
cell_clusters <- Idents(ALL.combined2)

# Count the number of cells in each cluster
cluster_counts <- table(cell_clusters)

# Print the results
print(cluster_counts)

# Convert the table to a data frame
cluster_counts_df <- as.data.frame(cluster_counts)
colnames(cluster_counts_df) <- c("Cluster", "Cell_Count")

# View the data frame
print(cluster_counts_df)


```{r}
# Load required libraries
if (!requireNamespace("ggplot2", quietly = TRUE)) {
    install.packages("ggplot2")
}
library(ggplot2)

library(Seurat)
library(dplyr)
library(Seurat)
library(dplyr)

# Load necessary libraries
library(dplyr)
library(ggplot2)
library(tidyr)
# Load necessary libraries
library(dplyr)
library(ggplot2)
library(tidyr)
library(dplyr)
library(ggplot2)
library(tidyr)

# Function to create sample DEG data frames for a specific cluster
# Load necessary libraries
library(ggplot2)
library(dplyr)

# Create mock data for all 16 clusters (replace these with your actual top gene data frames)
# Add the gene names as a new 
#0
top20cluster0_genes$gene <- rownames(top20cluster0_genes)

# Rearrange the columns to have gene names first (optional)
top20cluster0_genes <- top20cluster0_genes[, c("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "celltype")]
# Add the gene names as a new column
#1
top20cluster1_genes$gene <- rownames(top20cluster1_genes)
# Rearrange the columns to have gene names first (optional)
top20cluster1_genes <- top20cluster1_genes[, c("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "celltype")]
#2
# Add the gene names as a new column
top20cluster2_genes$gene <- rownames(top20cluster2_genes)

# Rearrange the columns to have gene names first (optional)
top20cluster2_genes <- top20cluster2_genes[, c("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "celltype")]
#3
# Add the gene names as a new column
top20cluster3_genes$gene <- rownames(top20cluster3_genes)

# Rearrange the columns to have gene names first (optional)
top20cluster3_genes <- top20cluster3_genes[, c("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "celltype")]
#4
# Add the gene names as a new column
top20cluster4_genes$gene <- rownames(top20cluster4_genes)

# Rearrange the columns to have gene names first (optional)
top20cluster4_genes <- top20cluster4_genes[, c("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "celltype")]
#5
top20cluster5_genes$gene <- rownames(top20cluster5_genes)
# Rearrange the columns to have gene names first (optional)
top20cluster5_genes <- top20cluster5_genes[, c("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "celltype")]
#6
# Rearrange the columns to have gene names first (optional)
top20cluster6_genes$gene <- rownames(top20cluster6_genes)

top20cluster6_genes <- top20cluster6_genes[, c("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "celltype")]
#7
top20cluster7_genes$gene <- rownames(top20cluster7_genes)

top20cluster7_genes <- top20cluster7_genes[, c("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "celltype")]
#8
top20cluster8_genes$gene <- rownames(top20cluster8_genes)

top20cluster8_genes <- top20cluster8_genes[, c("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "celltype")]
#9
top20cluster9_genes$gene <- rownames(top20cluster9_genes)
top20cluster9_genes <- top20cluster9_genes[, c("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "celltype")]
#10
top20cluster10_genes$gene <- rownames(top20cluster10_genes)
top20cluster10_genes <- top20cluster10_genes[, c("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "celltype")]
#11
top20cluster11_genes$gene <- rownames(top20cluster11_genes)
top20cluster11_genes <- top20cluster11_genes[, c("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "celltype")]
#12
top20cluster12_genes$gene <- rownames(top20cluster12_genes)

top20cluster12_genes <- top20cluster12_genes[, c("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "celltype")]
#13
top20cluster13_genes$gene <- rownames(top20cluster13_genes)

top20cluster13_genes <- top20cluster13_genes[, c("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "celltype")]
#14
top20cluster14_genes$gene <- rownames(top20cluster14_genes)

top20cluster14_genes <- top20cluster14_genes[, c("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "celltype")]
#15
top20cluster15_genes$gene <- rownames(top20cluster15_genes)

top20cluster15_genes <- top20cluster15_genes[, c("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "celltype")]

# Combine all top gene data frames into a list
all_top_genes <- list(
  Cluster0 = top20cluster0_genes,
  Cluster1 = top20cluster1_genes,
  Cluster2 = top20cluster2_genes,
  Cluster3 = top20cluster3_genes,
  Cluster4 = top20cluster4_genes,
  Cluster5 = top20cluster5_genes,
  Cluster6 = top20cluster6_genes,
  Cluster7 = top20cluster7_genes,
  Cluster8 = top20cluster8_genes,
  Cluster9 = top20cluster9_genes,
  Cluster10 = top20cluster10_genes,
  Cluster11 = top20cluster11_genes,
  Cluster12 = top20cluster12_genes,
  Cluster13 = top20cluster13_genes,
  Cluster14 = top20cluster14_genes,
  Cluster15 = top20cluster15_genes
)

# Combine all data frames into one
combined_top_genes <- bind_rows(all_top_genes, .id = "Cluster")
#combined_top_genes2 <- bind_rows(all_top_genes)

# Reset row names (optional)
rownames(combined_top_genes) <- NULL

# View combined data (optional)
print(combined_top_genes)

# Order genes for consistent plotting
combined_top_genes$gene <- factor(combined_top_genes$gene, 
                                   levels = unique(combined_top_genes$gene))

# Plotting
ggplot(combined_top_genes, aes(x = gene, y = avg_log2FC, fill = celltype)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  coord_flip() +  # Flip coordinates for better readability
  labs(title = "Top 20 DEGs Across 16 Clusters",
       x = "Genes",
       y = "Average Log2 Fold Change") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 5), 
        legend.title = element_blank())
library(dplyr)
############3
library(dplyr)
library(Seurat)

library(ggplot2)

# Assuming `combined_top_genes` is a data frame with columns:
# `gene`, `Cluster`, and `avg_log2FC`

ggplot(combined_top_genes, aes(x = celltype, y = gene, fill = avg_log2FC)) +
  geom_tile(color = "white") +  # Create heatmap tiles with a border
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 1) +  # Adjust colors
  labs(title = "Top 20 DEGs Across 16 Clusters",
       x = "celltype",
       y = "Genes",
       fill = "Avg Log2FC") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 5),
        axis.text.x = element_text(size = 10, angle = 45, hjust = 1),
        legend.title = element_blank())


##############
# Assuming combined_top_genes is your data frame and it contains 'gene', 'Cluster', and 'avg_log2FC'
install.packages("scales")
library(scales)

# Rescale avg_log2FC from -1 to 2
combined_top_genes$scaled_avg_log2FC <- rescale(combined_top_genes$avg_log2FC, to = c(-1, 2))

# Rescale avg_log2FC from -1 to 2
combined_top_genes$scaled_avg_log2FC <- rescale(combined_top_genes$avg_log2FC, to = c(-1, 2))

# Create the heatmap
ggplot(combined_top_genes, aes(x = celltype, y = gene, fill = scaled_avg_log2FC)) +
  geom_tile(color = "white") +  # Create heatmap tiles with a border
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0.5) +  # Adjust colors; midpoint at 0.5
  labs(title = "Top 20 DEGs Across 16 Clusters",
       x = "Cell Type",
       y = "Genes",
       fill = "Avg Log2FC") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 5),
        axis.text.x = element_text(size = 10, angle = 45, hjust = 1),
 
               legend.title = element_blank())
#######################

######################################################################
# Initialize a list to store DEGs for each cluster



```


```{r}
# Load required libraries
if (!requireNamespace("ggplot2", quietly = TRUE)) {
    install.packages("ggplot2")
}
library(ggplot2)

library(Seurat)
library(dplyr)
# Extract the top 10 DEGs based on adjusted p-value and ploting Dotplot and Heatmap
top20_deg_cluster0 <- head(cluster0_deg[order(cluster0_deg$p_val_adj), ], 20)
top20_genes <- rownames(top20_deg_cluster0)

# Create the dot plot for the top 10 DEGs
DotPlot(ALL.combined2, features = top10_genes, group.by = "orig.ident") +
  scale_color_gradientn(colors = c("blue", "green", "red")) +  # Customize color gradient
  labs(title = "Dot Plot of Top 20 DEGs: PDAC vs NOR CD8T Cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for readability

# Ensure data is scaled
#ALL.combined2 <- ScaleData(ALL.combined2, features = rownames(ALL.combined2))

# Create the heatmap for the top 10 DEGs
#DoHeatmap(ALL.combined2, features = top10_genes, group.by = "orig.ident") +
#  scale_fill_gradientn(colors = c("blue", "white", "red")) +
 # labs(title = "Heatmap of Top 10 DEGs: PDAC vs NOR CD8T Cells") +
 # theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for better #readability
##### FOR CLUSTER 1
# Extract the top 10 DEGs based on adjusted p-value and ploting Dotplot and Heatmap
top10_deg_cluster1 <- head(cluster1_deg[order(cluster1_deg$p_val_adj), ], 10)
top10_genes1 <- rownames(top10_deg_cluster1)

# Create the dot plot for the top 10 DEGs
DotPlot(ALL.combined2, features = top10_genes1, group.by = "orig.ident") +
  scale_color_gradientn(colors = c("blue", "green", "red")) +  # Customize color gradient
  labs(title = "Dot Plot of Top 10 DEGs: PDAC vs NOR naive B Cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for readability
# Ensure data is scaled
#ALL.combined2 <- ScaleData(ALL.combined2, features = rownames(ALL.combined2))
## CLUSTER 2
# Extract the top 10 DEGs based on adjusted p-value and ploting Dotplot and Heatmap
top10_deg_cluster2 <- head(cluster3_deg[order(cluster2_deg$p_val_adj), ], 10)
top10_genes2 <- rownames(top10_deg_cluster2)

# Create the dot plot for the top 10 DEGs
DotPlot(ALL.combined2, features = top10_genes2, group.by = "orig.ident") +
  scale_color_gradientn(colors = c("blue", "green", "red")) +  # Customize color gradient
  labs(title = "Dot Plot of Top 10 DEGs: PDAC vs NOR Epithelial Cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for readability
## CLUSTER 3
# Extract the top 10 DEGs based on adjusted p-value and ploting Dotplot and Heatmap
top10_deg_cluster3 <- head(cluster3_deg[order(cluster3_deg$p_val_adj), ], 10)
top10_genes3 <- rownames(top10_deg_cluster3)

# Create the dot plot for the top 10 DEGs
DotPlot(ALL.combined2, features = top10_genes3, group.by = "orig.ident") +
  scale_color_gradientn(colors = c("blue", "green", "red")) +  # Customize color gradient
  labs(title = "Dot Plot of Top 10 DEGs: PDAC vs NOR Fibroblast  Cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for readability
## CLUSTER 4
# Extract the top 10 DEGs based on adjusted p-value and ploting Dotplot and Heatmap
top10_deg_cluster4 <- head(cluster4_deg[order(cluster4_deg$p_val_adj), ], 10)
top10_genes4 <- rownames(top10_deg_cluster4)

# Create the dot plot for the top 10 DEGs
DotPlot(ALL.combined2, features = top10_genes4, group.by = "orig.ident") +
  scale_color_gradientn(colors = c("blue", "green", "red")) +  # Customize color gradient
  labs(title = "Dot Plot of Top 10 DEGs: PDAC vs NOR SMC Cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for readability
## CLUSTER 5
# Extract the top 10 DEGs based on adjusted p-value and ploting Dotplot and Heatmap
top10_deg_cluster5 <- head(cluster5_deg[order(cluster5_deg$p_val_adj), ], 10)
top10_genes5 <- rownames(top10_deg_cluster5)

# Create the dot plot for the top 10 DEGs
DotPlot(ALL.combined2, features = top10_genes5, group.by = "orig.ident") +
  scale_color_gradientn(colors = c("blue", "green", "red")) +  # Customize color gradient
  labs(title = "Dot Plot of Top 10 DEGs: PDAC vs NOR Macrophages Cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for readability
## CLUSTER 6
# Extract the top 10 DEGs based on adjusted p-value and ploting Dotplot and Heatmap
top10_deg_cluster6 <- head(cluster6_deg[order(cluster6_deg$p_val_adj), ], 10)
top10_genes6 <- rownames(top10_deg_cluster6)

# Create the dot plot for the top 10 DEGs
DotPlot(ALL.combined2, features = top10_genes6, group.by = "orig.ident") +
  scale_color_gradientn(colors = c("blue", "green", "red")) +  # Customize color gradient
  labs(title = "Dot Plot of Top 10 DEGs: PDAC vs NOR t-reg Cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for readability
## CLUSTER 7
# Extract the top 10 DEGs based on adjusted p-value and ploting Dotplot and Heatmap
top10_deg_cluster7 <- head(cluster7_deg[order(cluster7_deg$p_val_adj), ], 10)
top10_genes7 <- rownames(top10_deg_cluster7)

# Create the dot plot for the top 10 DEGs
DotPlot(ALL.combined2, features = top10_genes7, group.by = "orig.ident") +
  scale_color_gradientn(colors = c("blue", "green", "red")) +  # Customize color gradient
  labs(title = "Dot Plot of Top 10 DEGs: PDAC vs NOR Endothelial Cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for readability
## CLUSTER 8
# Extract the top 10 DEGs based on adjusted p-value and ploting Dotplot and Heatmap
top10_deg_cluster8 <- head(cluster8_deg[order(cluster8_deg$p_val_adj), ], 10)
top10_genes8 <- rownames(top10_deg_cluster8)

# Create the dot plot for the top 10 DEGs
DotPlot(ALL.combined2, features = top10_genes8, group.by = "orig.ident") +
  scale_color_gradientn(colors = c("blue", "green", "red")) +  # Customize color gradient
  labs(title = "Dot Plot of Top 10 DEGs: PDAC vs NOR Neutrophil Cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for readability
## CLUSTER 9
# Extract the top 10 DEGs based on adjusted p-value and ploting Dotplot and Heatmap
top10_deg_cluster9 <- head(cluster9_deg[order(cluster9_deg$p_val_adj), ], 10)
top10_genes9 <- rownames(top10_deg_cluster9)

# Create the dot plot for the top 10 DEGs
DotPlot(ALL.combined2, features = top10_genes9, group.by = "orig.ident") +
  scale_color_gradientn(colors = c("blue", "green", "red")) +  # Customize color gradient
  labs(title = "Dot Plot of Top 10 DEGs: PDAC vs NOR CD4T Cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for readability
## CLUSTER 10
# Extract the top 10 DEGs based on adjusted p-value and ploting Dotplot and Heatmap
top10_deg_cluster10 <- head(cluster10_deg[order(cluster10_deg$p_val_adj), ], 10)
top10_genes10 <- rownames(top10_deg_cluster10)

# Create the dot plot for the top 10 DEGs
DotPlot(ALL.combined2, features = top10_genes10, group.by = "orig.ident") +
  scale_color_gradientn(colors = c("blue", "green", "red")) +  # Customize color gradient
  labs(title = "Dot Plot of Top 10 DEGs: PDAC vs NOR NK Cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for readability
## CLUSTER 11
# Extract the top 10 DEGs based on adjusted p-value and ploting Dotplot and Heatmap
top10_deg_cluster11 <- head(cluster1_deg[order(cluster11_deg$p_val_adj), ], 10)
top10_genes11 <- rownames(top10_deg_cluster11)

# Create the dot plot for the top 10 DEGs
DotPlot(ALL.combined2, features = top10_genes11, group.by = "orig.ident") +
  scale_color_gradientn(colors = c("blue", "green", "red")) +  # Customize color gradient
  labs(title = "Dot Plot of Top 10 DEGs: PDAC vs NOR Basophil Cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for readability
## CLUSTER 12
# Extract the top 10 DEGs based on adjusted p-value and ploting Dotplot and Heatmap
top10_deg_cluster12 <- head(cluster12_deg[order(cluster12_deg$p_val_adj), ], 10)
top10_genes12 <- rownames(top10_deg_cluster12)
# Create the dot plot for the top 10 DEGs
DotPlot(ALL.combined2, features = top10_genes12, group.by = "orig.ident") +
  scale_color_gradientn(colors = c("blue", "green", "red")) +  # Customize color gradient
  labs(title = "Dot Plot of Top 10 DEGs: PDAC vs NOR Oligodendrocytes Cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for readability
###In your color scale:
#Blue represents lower or negative expression levels.
#Green is a midpoint, representing moderate expression.
#Red indicates high expression levels.
## CLUSTER 13
# Extract the top 10 DEGs based on adjusted p-value and ploting Dotplot and Heatmap
top10_deg_cluster13 <- head(cluster13_deg[order(cluster13_deg$p_val_adj), ], 10)
top10_genes13 <- rownames(top10_deg_cluster13)

# Create the dot plot for the top 10 DEGs
DotPlot(ALL.combined2, features = top10_genes13, group.by = "orig.ident") +
  scale_color_gradientn(colors = c("blue", "green", "red")) +  # Customize color gradient
  labs(title = "Dot Plot of Top 10 DEGs: PDAC vs NOR B Cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for readability
## CLUSTER 14
# Extract the top 10 DEGs based on adjusted p-value and ploting Dotplot and Heatmap
top10_deg_cluster14 <- head(cluster14_deg[order(cluster14_deg$p_val_adj), ], 10)
top10_genes14 <- rownames(top10_deg_cluster14)

# Create the dot plot for the top 10 DEGs
DotPlot(ALL.combined2, features = top10_genes14, group.by = "orig.ident") +
  scale_color_gradientn(colors = c("blue", "green", "red")) +  # Customize color gradient
  labs(title = "Dot Plot of Top 10 DEGs: PDAC vs NOR naive T Cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for readability
## CLUSTER 15
# Extract the top 10 DEGs based on adjusted p-value and ploting Dotplot and Heatmap
top10_deg_cluster15 <- head(cluster15_deg[order(cluster15_deg$p_val_adj), ], 10)
top10_genes15 <- rownames(top10_deg_cluster15)

# Create the dot plot for the top 10 DEGs
DotPlot(ALL.combined2, features = top10_genes15, group.by = "orig.ident") +
  scale_color_gradientn(colors = c("blue", "green", "red")) +  # Customize color gradient
  labs(title = "Dot Plot of Top 10 DEGs: PDAC vs NOR Endocrine  Cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for readability

```
The "prevalence of gene expression across cells in the specified group" refers to how commonly or frequently a particular gene is expressed among the cells in a given group (like `"PDAC"` or `"NOR"` in your dataset).

In other words, **prevalence** indicates the **percentage of cells within each group** that have detectable expression of a particular gene.

### Breaking It Down

Let's take an example to clarify:

1. **Gene Expression in Single-Cell RNA-seq**: In single-cell RNA sequencing, each gene's expression level can vary widely across individual cells, even within the same type of cell. Some genes may be expressed in many cells, while others might be expressed in only a few cells.

2. **Dot Plot Dot Size as a Measure of Prevalence**:
   - In the dot plot, the **dot size** for each gene represents the **percentage of cells in that group** where the gene's expression is detectable above a certain threshold.
   - A **larger dot** means that a **higher percentage of cells** in that group express the gene.
   - A **smaller dot** means that **fewer cells** in the group express the gene.

### Example to Illustrate Prevalence

Suppose you have two groups of cells:
- **PDAC** (cancerous cells)
- **NOR** (normal cells)

And you’re looking at the gene `FOS`:
- If `FOS` has a **large dot** in the `PDAC` group, it means that `FOS` is expressed in a large percentage of cells in the `PDAC` group.
- If `FOS` has a **small dot** in the `NOR` group, it indicates that only a small percentage of cells in the `NOR` group express `FOS`.

This shows that `FOS` is **more prevalent** in the `PDAC` group than in the `NOR` group, meaning that `FOS` is **expressed more widely** across the cancer cells than in the normal cells.

### Summary

- **Prevalence** in this context answers the question: "How many cells in this group express this gene?"
- It helps you understand whether a gene is commonly or rarely expressed within each cell group.
- This information, combined with the **color** (which shows the **average expression level**), gives a full picture of how a gene behaves in different cell groups.
#####################################################################################


```{r}

# Load clusterProfiler for enrichment analysis
library(clusterProfiler)
library(org.Hs.eg.db) # Human database for gene enrichment
library(clusterProfiler)
library(org.Hs.eg.db)  # For human gene annotation
library(dplyr)
#install.packages("tidyr")
library(Seurat)
library(dplyr)
Idents(ALL.combined2) <- "orig.ident"  # Make sure to set the appropriate metadata column for identities

# Perform differential expression analysis
Idents(ALL.combined2)
markers <- FindMarkers(ALL.combined2, ident.1 = "PDAC", ident.2 = "NOR")
Idents(ALL.combined2) <- "orig.ident"  # Make sure to set the appropriate metadata column for identities
# Get upregulated genes (positive log fold change)
upregulated_genes <- markers %>% filter(avg_log2FC > 0)

# Get downregulated genes (negative log fold change)
downregulated_genes <- markers %>% filter(avg_log2FC < 0)

# Check the results
head(markers)

# Filter and get the top 50 DEGs based on p_val_adj
top_50_deg <- markers %>%
  as.data.frame() %>%  # Ensure it's a dataframe for easier manipulation
  arrange(p_val_adj) %>%  # Sort by adjusted p-value
  head(50)  # Get the top 50

# Optionally, you can also arrange by absolute log2FoldChange if desired
# Filter and get the top 50 DEGs based on p_val_adj, then sort by p-value in descending order
top_50_deg <- markers %>%
  as.data.frame() %>%  # Ensure it's a dataframe for easier manipulation
  arrange(p_val_adj) %>%  # Sort by adjusted p-value in ascending order
  head(50) %>%  # Take the top 50 entries with the smallest p_val_adj
  arrange(desc(p_val_adj))  # Sort those top 50 by p_val_adj in descending order

# top_50_deg <- markers %>%
#   as.data.frame() %>%
#   arrange(desc(abs(log2FoldChange))) %>%
#   head(50)
# Filter and get the top 80 DEGs based on p_val_adj, then sort by p-value in descending order
#top_80_deg <- markers %>%
  as.data.frame() %>%  # Ensure it's a dataframe for easier manipulation
  arrange(p_val_adj) %>%  # Sort by adjusted p-value in ascending order
  head(80) %>%  # Take the top 80 entries with the smallest p_val_adj
  arrange(desc(p_val_adj))  
  #############3# Sort those top 80 by p_val_adj in descending order
top_80_deg2 <- markers %>%
  as.data.frame() %>%  # Ensure it's a dataframe for easier manipulation
  filter(p_val_adj < 0.01) %>%  # Only keep rows with p_val_adj < 0.01
  arrange(p_val_adj) %>%  # Sort by adjusted p-value in ascending order
  head(80) %>%  # Take the top 80 entries with the smallest p_val_adj
  arrange(desc(p_val_adj))  # Sort those top 80 by p_val_adj in descending order

# View the top 50 DEGs
#print(top_50_deg)

# Example: Save the top 50 DEGs to a CSV file
#write.csv(top_50_deg, "top_50_DEGs_PD_vs_NOR.csv")

# You can visualize the top DEGs using a heatmap, for example:
# Here we assume you have a function to get expression data
top_genes <- rownames(top_80_deg2)

# Assuming 'top_50_deg' is your filtered markers dataframe
top_genes <- rownames(top_80_deg2)  # Extract the gene names

# Convert gene symbols to ENTREZ IDs
gene_ids <- bitr(top_genes, fromType = "SYMBOL", toType = "ENTREZID", OrgDb = org.Hs.eg.db)

# Check conversion results
head(gene_ids)
##
# Perform GO enrichment analysis
ego <- enrichGO(gene = gene_ids$ENTREZID,
                OrgDb = org.Hs.eg.db,
                keyType = "ENTREZID",
                ont = "BP",  # Biological Process; can also use "MF" or "CC"
                pAdjustMethod = "BH",
                qvalueCutoff = 0.05)

# View the enrichment results
head(ego)
# Dotplot for GO enrichment results
#dotplot(ego, showCategory = 50) 
# Correct usage of dotplot for GO enrichment results
library(clusterProfiler)
library(org.Hs.eg.db)  # For human gene identifiers
library(dplyr)
library(tidyr)

dotplot(ego, showCategory = 10) 
# Optionally, perform KEGG enrichment analysis
ekegg <- enrichKEGG(gene = gene_ids$ENTREZID,
                    organism = 'hsa',  # For human
                    pvalueCutoff = 0.05)

# View the KEGG enrichment results
head(ekegg)

# Visualize KEGG results
dotplot(ekegg, showCategory = 10) 
dotplot(ego, showCategory = 10) 
barplot(ego, showCategory = 10)
barplot(ekegg,showCategory = 10)
```







